Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,40 +29,83 @@
 #include "TextureMapperLayer3DRenderingContext.h"
 
 #include "ClipPath.h"
+#include "FloatBox3D.h"
 #include "FloatPlane3D.h"
+#include "FloatPolygon3D.h"
+#include "GeometryUtilities.h"
 #include "TextureMapperGPUBuffer.h"
 #include "TextureMapperLayer.h"
-#include <wtf/Deque.h>
+#include <numeric>
+#include <wtf/StdSet.h>
 #include <wtf/TZoneMallocInlines.h>
 
 namespace WebCore {
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(TextureMapperLayer3DRenderingContext);
 
-void TextureMapperLayer3DRenderingContext::paint(TextureMapper& textureMapper, const Vector<TextureMapperLayer*>& layers,
+void TextureMapperLayer3DRenderingContext::paint(TextureMapper& textureMapper, const Vector<TextureMapperLayer*>& textureMapperLayers,
     const std::function<void(TextureMapperLayer*, const ClipPath&)>& paintLayerFunction)
 {
-    if (layers.isEmpty())
+    if (textureMapperLayers.isEmpty())
+        return;
+
+    Vector<Layer> layers;
+    for (auto* layer : textureMapperLayers)
+        layers.append({ { layer->effectiveLayerRect(), layer->toSurfaceTransform() }, layer, false });
```

Comment:
That is hard to read -- one can only guess what objects you are constructing here. The spurious boolean in the end raises eyebrows. `layers.append(Layer::createFoo(layer))` , and have  static createFoo method encapsulating the logic in `Layer`?

Reply:
Didn't add factory function but changed it to be more readable.